### ISSUE TITLE
inference on new data samples

### ISSUE THREAD

**Issue by sophark** [kLabUM/rrcf#70]
Hi, Thanks for implementing this. I have a use case that need train the rrcf using some given dataset, and then predict on unseen data samples this haven't been seen during training period. 

Can we use batch mode to achieve that? One simple solution I can come up with is to first insert that point to the forest, calculate the codisp, and then delete it. I am wondering is there any smarter ways to save the inference time?

Thanks.

**Comment by mdbartos** [comment]
That's probably the most flexible way to do it (create forest from point set S using batch mode -> insert new point x into each tree -> compute codisp -> delete point x). But yes, it will probably be slow. Parallelizing can help though.

I'm not sure if this is helpful, but note that the insert_point algorithm is guaranteed to produce a tree drawn from S \union x (where S is a point set, and x is an additional point).

In other words, the following two trees are statistically indistinguishable, and the codisp of x will be the same in expectation:

- Create tree T' from point set (S \union x) via batch mode.
- Create tree T from point set S via batch mode and then insert x, resulting in tree T'.

**Comment by sophark** [comment]
> That's probably the most flexible way to do it (create forest from point set S using batch mode -> insert new point x into each tree -> compute codisp -> delete point x). But yes, it will probably be slow. Parallelizing can help though.

Thanks for your hints. Yes, it indeed a little bit slow without parallelizing. Do you know which step above consume most of time and its time complexity? I guess maybe the insert point step?

**Comment by mdbartos** [comment]
Yeah, I would say insert_point is the slowest step. I have time breakdowns here:
https://github.com/kLabUM/rrcf/issues/28
